### DISCUSSION TITLE
How can an msc storage device detach itself?

### DISCUSSION THREAD

**Discussion by tlokovic** [hathach/tinyusb#1469] [General]
I think I understand how an msc storage device can react when the host tells it "you're ejected", but how can the device itself say it's going away? 

In this example:

https://github.com/hathach/tinyusb/blob/master/examples/device/msc_dual_lun/src/msc_disk_dual.c

I see

// Invoked when received Test Unit Ready command.
// return true allowing host to read/write this LUN e.g SD card inserted
bool tud_msc_test_unit_ready_cb(uint8_t lun)
{
  if ( lun == 1 && board_button_read() ) return false;

  return true; // RAM disk is always ready
}

But triggering this doesn't actually seem to tell Windows that the device has ejected.  It just makes reads fail while the button is pressed.  What I'd like is for the device to disappear when I push the button and then reappear when I release the button.

Is there a way for me to initiate the removal and re-addition of msc storage devices?

Thank you!

**Comment by tlokovic** [comment]
Ah, I think the answer may have been given in https://github.com/hathach/tinyusb/discussions/584:

If tud_msc_test_unit_ready_cb() calls

tud_msc_set_sense(lun, SCSI_SENSE_NOT_READY, 0x3a, 0x00);

whenever it's about to return false, that seems to tell the host that the LUN is actually gone.

In Windows, this makes the mounted drive disappear until  tud_msc_test_unit_ready_cb() returns true, at which point the mounted drive reappears.  This is exactly the behavior I wanted.  Interestingly, it appears and disappears without playing the USB-attach/detach sounds, because I guess the USB device is there the whole time, reporting the same number of LUNs all the time, but just declaring some of them ready and some of them not.  I didn't realize that was a thing.

So now I can have the LUN say "I'm outta here" and then "Hey I'm back!".  I'm still curious if there's a way in TinyUSB for the USB device itself to say "I'm outta here" and then "Hey, I'm back!"

**Comment by smunaut** [comment]
No, tinyUSB doesn't have a way to trigger a disconnect.

**Reply by hathach** [reply]
tud_disconnect() can be used to trigger disconnection by disabling pull-up register on data line for supported MCUs.

**Reply by smunaut** [reply]
Huh, never mind my bad ... I thought I had looked for that a long time ago and didn't find it ...

**Reply by hathach** [reply]
no problem at all, you probably right previously, tinyusb is still evolving :)

**Comment by hathach** [comment]
@tlokovic I am not entirely sure what you want to achieve. Can you confirm which of following cases :
1. device fully disconnected, which require full re-enumeration essentially like you pull the device off
2. similar to remove SD card from its card reader

**Reply by tlokovic** [reply]
@hathach you're right, I didn't make my goal clear.

My goal is to create a device that has some non-persistent storage (basically a ramdisk) which can be accessed from two different computers, one seeing the data as a FAT fileysystem over USB, and the other (a Commodore 64) seeing the data as an emulated floppy drive (through the Commodore IEC serial bus).  The two systems won't be able to access the data at the same time: a physical switch will indicate which system has control of the data at any given time.

For the USB side, I have a Pi Pico which (thanks to TinyUSB and what I learned from msc_disk_dual.c) is presenting itself over USB as a FAT12 filesystem.  So far so good.  But when I want to switch away from the USB view, I need a way to tell the USB side "it's not your turn any more", and have the drive be unmounted.  Later, when I switch it back to the USB side, I want it to re-mount and refresh its view of the data in case it was changed by the other system.

My problem seems to be solved by tud_msc_set_sense(lun, SCSI_SENSE_NOT_READY, 0x3a, 0x00).   At least on Windows, the mounted drive disappears when I call that, and reappears when I start returning ready=true again.  So that's great.

Maybe this is what you mean by "similar to remove SD card from its card reader", though the SD Card reader I own plays the USB connect/disconnect sounds when I insert and remove SD cards.  SCSI_SENSE_NOT_READY does not.

So I think I'm all set for now, unless you have reason to believe I should be unmounting/remounting differently than I am.  To your question, I don't know which kind of "disconnect" is the right one to do here.  I wasn't sure if it was considered bad form for a USB device to report "not ready" on one of its LUNs for long stretches of time.  Is that a reasonable thing for a device to do, or is that considered more of an error condition?  It seems to work in practice, but I don't know if it's "bad" for some reason unknown to me.

**Reply by hathach** [reply]
`tud_msc_set_sense(lun, SCSI_SENSE_NOT_READY, 0x3a, 0x00)` is the standard behavior for SD card reader. Some of them will **self-unplug** to save power if there is no card at all, however if you have e.g 2 cards inserted than remove 1, you could observe the LUN where the card is removed reported the above sense. I am glad you figure it out.

**Reply by tlokovic** [reply]
Thank you for the response, and for your work on this amazing library!
